NDiff

petenanopay opened the pull request.

Commit:
NDiff: start roughing out (NOT WORKING)

Commit:
NDiff: get thing sort of working

Commit:
Merge branch 'development' into feature/NDiff

jlhughes commented on src/foam/nanos/ndiff/NDiffDAO.js:
```diff
@@ -0,0 +1,72 @@
+            setDelegate(delegate);
+
+            foam.nanos.logger.Logger logger = (foam.nanos.logger.Logger) getX().get("logger");
+            logger.info("NDiffDAO: object created");
```
This can be simplified with 
Loggers logger = Loggers.logger(x, this);
this will set up PrefixLogger with the class name, then you don't need to add "NDiffDAO" to each log call. 
Also, use  commas ',' rather than : and concatenation. 
logger.info("constructor");

petenanopay replied:
now using the Loggers API to get a logger

jlhughes commented on src/foam/nanos/ndiff/NDiffDAO.js:
```diff
@@ -0,0 +1,72 @@
+            // WRONG! can't grab nspec name at the moment!
+            // only temporary, will change soon
+            String nSpecName = obj.getClass().getName();
+            String dbg = "(nSpecName="+nSpecName+",objectId="+objectId+")";
```
continuing on logging, don't to the concatenation. it will occur only if debug is enabled, for example. And we are good with , separating key/value pairs.
so assuming logger setup with Loggers
logger.info("nSpecName", nSpecName, "objectId", objectid, "find", ...);

jlhughes commented on src/foam/nanos/ndiff/NDiffDAO.js:
```diff
@@ -0,0 +1,72 @@
+            NDiff ndiff = existingNdiff != null ? existingNdiff : new NDiff();
+            ndiff.setObjectId(objectId);
+            ndiff.setNSpecName(nSpecName); 
+            if (true) { 
```
what is this?
assuming so you flip during your testing.

jlhughes commented on src/foam/nanos/ndiff/NDiffDAO.js:
```diff
@@ -0,0 +1,72 @@
+                ndiff.setRuntimeFObject(obj);
+            }
+
+            super.put_(x, ndiff); 
```
```suggestion
            return getDelegate().put_(x, obj); 
```

petenanopay replied:
missed this earlier; now changed

jlhughes commented on src/foam/nanos/ndiff/NDiffDAO.js:
```diff
@@ -0,0 +1,72 @@
+            foam.nanos.logger.Logger logger = (foam.nanos.logger.Logger) getX().get("logger");
+
+            if ( obj instanceof NDiff ) {
+                return super.put_( x, obj );
```
When does this occur?

petenanopay replied:
long since changed

jlhughes commented on src/foam/nanos/ndiff/NDiffDAO.js:
```diff
@@ -0,0 +1,72 @@
+            // only temporary, will change soon
+            String nSpecName = obj.getClass().getName();
+            String dbg = "(nSpecName="+nSpecName+",objectId="+objectId+")";
+
```
DAO ndiffDAO = (DAO) x.get("ndiffDAO");

petenanopay replied:
code now grabs ndiffDAO service and uses it

jlhughes commented on src/foam/nanos/ndiff/NDiffDAO.js:
```diff
@@ -0,0 +1,72 @@
+            String dbg = "(nSpecName="+nSpecName+",objectId="+objectId+")";
+
+            logger.info("NDiffDAO: try find existing ndiff: "+dbg);
+            NDiff existingNdiff = (NDiff) super.find(
```
```suggestion
            NDiff existingNdiff = (NDiff) ndiffDAO.find(
```

jlhughes commented on src/foam/nanos/ndiff/NDiffDAO.js:
```diff
@@ -0,0 +1,72 @@
+                ndiff.setRuntimeFObject(obj);
+            }
+
+            super.put_(x, ndiff); 
```
```suggestion
            ndiffDAO.put_(x, ndiff); 
```

jlhughes commented on src/foam/nanos/ndiff/services.jrl:
```diff
@@ -0,0 +1,22 @@
+  "serviceScript": """
+    dao  = new foam.dao.EasyDAO.Builder(x)
+      .setOf(foam.nanos.ndiff.NDiff.getOwnClassInfo())
+      .setDecorator(new foam.nanos.ndiff.NDiffDAO(x, new foam.dao.NullDAO(x, foam.nanos.ndiff.NDiff.getOwnClassInfo())))
```
Ok, couple of options
1. this is not required. and NDiffDAO created in JDAO looksup 'ndiffDAO' as needed. 
2. the composite dao just setups two daos    x.get("ndiffDAO"), and the original delegate. and puts to each.  This decorator here will handle the put on the ndiffdao and do the appropriate adapting, and doesn't have to reply or delegate itself.

petenanopay replied:
the JRL file has changed; we now grab the ndiffDAO service in the NDiffDAO decorator.

jlhughes reviewed (commented):
Lots of comments, suggest you start at the bottom and work up.

Commit:
NDiff: different pipeline to store ndiffs

Commit:
merge development branch

Commit:
NDiff: codestyle cleanups, kill CompositeDAO

Commit:
NDiff: small codereview change

Commit:
NDiff: optimization: avoid FUID generation

Commit:
NDiff: actually store by nspec/decorate easydao

jlhughes commented on src/foam/nanos/boot/NSpec.js:
```diff
@@ -11,6 +11,17 @@ foam.CLASS({
 
+  constants: [
+    {
+      name: 'NSPEC_NAME_CTX_KEY',
```
```suggestion
      name: 'NSPEC_CTX_KEY',
```
Since you are passing the entire NSPEC.

jlhughes commented on src/foam/nanos/ndiff/NDiffDAO.js:
```diff
@@ -0,0 +1,92 @@
+            Object objectId = storedObject.getProperty("id");
+            String nSpecName = getNSpecName();
+            
+            NDiff existingNdiff = (NDiff) ndiffDao.find(
```
Since you've added
ids: [property1, property2]  to NDiff, you can now use the 
NDiffID FObject for your find. , look in your build/src/java for it.
Only if you think it helps.

Commit:
merge development branch

Commit:
NDiff: code review changes

Commit:
NDiff: two small changes

Commit:
Merge remote-tracking branch 'origin/development' into feature/NDiff

Commit:
Merge remote-tracking branch 'origin/development' into feature/NDiff

Commit:
Merge remote-tracking branch 'origin/development' into feature/NDiff

Commit:
Ndiff: remove authenticate:true, fixes tests

Commit:
NDiff: add menus.jrl

Commit:
NDiff: change objectId to a String

Commit:
NDiff: monitor deletions, update delta during select

at the moment the delta flag updates are happening after filters and
sorting are being applied. this is causing issues in the views. fixing
this will be a do-later

Commit:
Merge remote-tracking branch 'origin/development' into feature/NDiff

Commit:
NDiff: add Apply Original function

jlhughes commented on src/foam/dao/EasyDAO.js:
```diff
@@ -193,6 +196,20 @@ foam.CLASS({
+        // (this should ONLY be enabled for nspecs that have journals!)
+        if ( getNSpec() != null &&
+             getJournalType().equals(JournalType.SINGLE_JOURNAL) &&
+             !getWriteOnly() &&
```
```suggestion
             ! getWriteOnly() &&
```

jlhughes commented on src/foam/dao/EasyDAO.js:
```diff
@@ -836,9 +853,9 @@ model from which to test ServiceProvider ID (spid)`,
         if ( getJournalType().equals(JournalType.SINGLE_JOURNAL) ) {
           if ( getWriteOnly() ) {
-            delegate = new foam.dao.WriteOnlyJDAO(x, delegate, getOf(), getJournalName());
+            delegate = new foam.dao.WriteOnlyJDAO(subx, delegate, getOf(), getJournalName());
```
subx is never defined

petenanopay replied:
as discussed, this was refinement-related weirdness. changed back to "x"

jlhughes commented on src/foam/nanos/ndiff/NDiff.js:
```diff
@@ -0,0 +1,75 @@
+    },
+    {
+        class: 'Boolean',
+        name: 'applyingOriginal',
```
```suggestion
        name: 'applyOriginal',
```

jlhughes commented on src/foam/nanos/ndiff/NDiffDAO.js:
```diff
@@ -0,0 +1,111 @@
+    if it's running. 
+    `,
+    javaImports: [
+        'foam.nanos.logger.Logger',
```
Can you set your editor to 2 spaces. Thanks.

petenanopay replied:
all ndiff sourcefiles converted to two space - further issues, if any, to be corrected down the line

jlhughes commented on src/foam/nanos/ndiff/NDiffDAO.js:
```diff
@@ -0,0 +1,111 @@
+
+            String objectId = storedObject.getProperty("id").toString();
+            String nSpecName = getNSpecName();
+            NDiff existingNdiff = (NDiff) ndiffDao.find(new NDiffId.Builder(x)
```
Use the generated constructor for creating the NDiffId, less garbage generated.

jlhughes commented on src/foam/nanos/ndiff/NDiffRuntimeDAO.js:
```diff
@@ -0,0 +1,94 @@
+            
+            Sink originalSink = prepareSink(sink);
+
+            // s2 exists because we need to return the original sink later
```
```suggestion
            // originalSink exists because we need to return the original sink later
```

Commit:
NDiff: code review changes

Commit:
Update src/foam/dao/EasyDAO.js

Co-authored-by: Joel Hughes <jlhughes@users.noreply.github.com>

Commit:
Merge development into branch

Commit:
NDiff: work in progress comparison view

Commit:
Merge remote-tracking branch 'origin/development' into feature/NDiff

Commit:
NDiff: better comparisonview, fix isEnabled checks

jlhughes commented on src/foam/dao/EasyDAO.js:
```diff
@@ -193,6 +196,20 @@ foam.CLASS({
+        // listen for further changes to journaled nspec entries
+        // and forward them to NDiffDAO
+        // (this should ONLY be enabled for nspecs that have journals!)
+        if ( getNSpec() != null &&
```
With the 'sink' for runtime changes, is this still necessary?

Commit:
NDiff: table widths

Commit:
NDiff: compare/store runtimeFObject on select
